compiler/rustc_passes/src/debugger_visualizer.rs

```diff
@@ -0,0 +1,133 @@
+                };
+
+                let file = match (meta_item.name_or_empty(), meta_item.value_str()) {
+                    (sym::natvis_file, Some(value)) => {
+                        match resolve_path(&self.tcx.sess.parse_sess, value.as_str(), attr.span) {
+                            Ok(file) => file,
+                            Err(mut err) => {
+                                err.emit();
+                                continue;
+                            }
+                        }
+                    }
+                    (_, _) => continue,
+                };
+
+                match (file.exists(), file.is_file()) {
+                    (true, true) => {
+                        let contents = match std::fs::read_to_string(&file) {
+                            Ok(contents) => contents,
+                            Err(err) => {
+                                self.tcx
+                                    .sess
+                                    .struct_span_err(
+                                        attr.span,
+                                        &format!(
+                                            "Unable to read contents of file `{}`. {}",
+                                            file.display(),
+                                            err
+                                        ),
+                                    )
+                                    .emit();
+                                continue;
+                            }
+                        };
+
+                        let file_name = FileName::from(file.clone());
+                        self.debugger_visualizers.insert(DebuggerVisualizerFile::new(
+                            file_name,
+                            contents,
+                            DebuggerVisualizerType::Natvis,
+                        ));
+                    }
+                    (_, _) => {
+                        self.tcx
+                            .sess
+                            .struct_span_err(
+                                attr.span,
+                                &format!("{} wasn't a valid file", file.display()),
+                            )
+                            .emit();
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Traverses and collects all the debugger visualizers in this crate.
+fn debugger_visualizers(tcx: TyCtxt<'_>, (): ()) -> Vec<DebuggerVisualizerFile> {
+    // Initialize the collector.
+    let mut collector = DebuggerVisualizerCollector::new(tcx);
+
+    // Collect debugger visualizers in this crate.
+    tcx.hir().visit_all_item_likes(&mut collector);
+
+    // Collect debugger visualizers on the crate attributes.
+    collector.check_for_debugger_visualizer(CRATE_HIR_ID);
+
+    // Extract out the found debugger_visualizer items.
+    let DebuggerVisualizerCollector { debugger_visualizers, .. } = collector;
+
+    debugger_visualizers.into_iter().collect::<Vec<_>>()
```

Comment:
The `Vec<>` returned here must be in a deterministic order (since this is a query provider). We can either sort the vector or make sure `DebuggerVisualizerCollector` produces a deterministic output.